src/source-control/GitSourceControl.ts

```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
+            });
+
+            // Stop if this is the common ancestor commit
+            if (nodegitCommit.id().equal(commonAncestorCommitOid)) {
+              stopped = true;
+            }
+          });
+
+          history.on("end", () => resolve());
+          history.start();
+        });
+      }),
+    );
+
+    // Build relationships
+    commitChildrenMap.forEach((childrenHashes, parentHash) => {
+      const parentCommit = hashMap.get(parentHash);
+      if (!parentCommit) {
+        return;
+      }
+      childrenHashes.forEach((childHash) => {
+        const childCommit: Commit = hashMap.get(childHash)!;
+        childCommit.parentCommits = Array.from(
+          new Set([...childCommit.parentCommits, parentHash]),
+        );
+        parentCommit.childCommits = Array.from(
+          new Set([...parentCommit.childCommits, childHash]),
+        );
+        hashMap.set(childHash, childCommit);
+      });
+      hashMap.set(parentHash, parentCommit);
+    });
+
+    // Inject branch names
+
+    refs.forEach((ref) => {
+      const branchName = ref.name();
+      const branchSha = ref.target().tostrS();
+      hashMap.get(branchSha)?.refNames.push(branchName);
+    });
+
+    const hasUncommittedChanges = repoStatus.length > 0;
+    // Earliest Interesting Commit must exist at this point
+    const earliestInterestingCommit = hashMap.get(
+      commonAncestorCommitOid.tostrS(),
+    )!;
+
+    const ourRepository: Repository = {
+      path: this.repoPath,
+      hasUncommittedChanges,
+      headHash,
+      earliestInterestingCommit,
+      commits: hashMap,
+    };
+
+    this.repo = ourRepository;
   }
 
   async getCommitByHash(hash: CommitHash): Promise<Commit | null> {
-    // TODO: Implement
-    return null;
+    // TODO: Implement for partial match
+    const relatedCommits: Commit[] = [];
+    const keys = Array.from(this.repo.commits.keys());
+    keys.forEach((key) => {
+      if (key.indexOf(hash) !== -1) {
+        // If hash in key, as the key was from the commits' key, it should be present
+        relatedCommits.push(this.repo.commits.get(key)!);
+      }
+    });
+    if (relatedCommits.length !== 1) {
+      return null;
+    } else {
+      return relatedCommits[0];
+    }
   }
 
   async rebaseCommits(
-    rebaseRootCommit: Commit,
-    targetCommit: Commit,
+    rebaseRootCommit: CommitHash,
+    targetCommit: CommitHash,
   ): Promise<void> {
-    // TODO: Implement
+    const repo = await nodegit.Repository.open(this.repoPath);
+    await this.loadIfChangesPresent();
+    const tempBranchName = `sttack-${lorem.slug(3)}`;
+    const queue: string[] = [rebaseRootCommit];
+    // 1. Rebase root commit on target commit
+    // 2. Update Target Commit to point to the cherry picked commit
+    // 3. Update rebaseRootCommit to point to child of itself
+    // Call LoadRepositoryInformation
+
+    while (queue.length) {
+      const hashToBeRebased = queue.pop();
+      // This rebase commit must exist
```

Comment:
This commit is a bit ambiguous - is this a prerequisite, an assumption, or a declaration of some fact?

Reply:
It's a prerequisite and also an assumption, because we populate our repo with commits prior to calling this function

Reply:
I'll throw an error if either of these does not exist, updating this.

Reply:
Yeah that should be okay. The frontend can prettify it eventually